diff_drive_controller/src/velocity_limiter.cpp

```diff
@@ -0,0 +1,154 @@
+        return;
+    }
+
+    if (_has_velocity_limits) {
+        limitVel(vel_cmd);
+    }
+
+    if (_has_acceleration_limits) {
+        limitAcc(vel_cmd, vel_cmd_prev, dt);
+    }
+    return;
+};
+
+void VelocityLimiter::limitVel(Vector& vel_cmd) const {
+    // initial down-scaling for angular velocity limit
+    downscaleBoth(vel_cmd.x, -_vel_x_max, _vel_x_max, vel_cmd.th, -_vel_th_max, _vel_th_max);
+
+    // convert to wheel velocities
+    float v_left = vel_cmd.x - vel_cmd.th * _wheel_separation * 0.5f;
+    float v_right = vel_cmd.x + vel_cmd.th * _wheel_separation * 0.5f;
+
+    // clip
+    downscaleBoth(v_left, v_right, -_vel_x_max, _vel_x_max);
+
+    // convert back to x/th velocities
+    vel_cmd.x = 0.5 * (v_left + v_right);
+    vel_cmd.th = (v_right - v_left) / _wheel_separation;
+}
+
+void VelocityLimiter::limitAcc(Vector& vel_cmd, const Vector& vel_cmd_prev, const float dt) const {
+    Vector acc{(vel_cmd.x - vel_cmd_prev.x) / dt, (vel_cmd.th - vel_cmd_prev.th) / dt};
+
+    // calculate wheel acceleration forces
+    float f_left = 0.5f * _mass * acc.x - _inertia * acc.th / _wheel_separation;
+    float f_right = 0.5f * _mass * acc.x + _inertia * acc.th / _wheel_separation;
+
+    // calculate previous wheel velocities
+    const float v_left_prev = vel_cmd_prev.x - vel_cmd_prev.th * _wheel_separation * 0.5f;
+    const float v_right_prev = vel_cmd_prev.x + vel_cmd_prev.th * _wheel_separation * 0.5f;
+
+    // force limits:
+    // - if wheel velocity is positive, acceleration is limited to within [_f_wheel_min, _f_wheel_max]
+    // - if wheel velocity is negative, acceleration is limited to within [-_f_wheel_max, -_f_wheel_min]
+    const float f_left_min = v_left_prev > 0 ? _f_wheel_min : -_f_wheel_max;
+    const float f_left_max = v_left_prev > 0 ? _f_wheel_max : -_f_wheel_min;
+    const float f_right_min = v_right_prev > 0 ? _f_wheel_min : -_f_wheel_max;
+    const float f_right_max = v_right_prev > 0 ? _f_wheel_max : -_f_wheel_min;
+
+    // limit
+    downscaleBoth(f_left, f_left_min, f_left_max, f_right, f_right_min, f_right_max);
+
+    // convert back to to x/th accelerations
+    acc.x = 1.0f / _mass * (f_left + f_right);
+    acc.th = 0.5f * _wheel_separation / _inertia * (f_right - f_left);
+
+    // convert back to vel
+    vel_cmd.x = vel_cmd_prev.x + acc.x * dt;
+    vel_cmd.th = vel_cmd_prev.th + acc.th * dt;
+}
+
+void VelocityLimiter::downscaleBoth(
+        float& v1, const float v1_min, const float v1_max, float& v2, const float v2_min, const float v2_max) const {
+    const float v1_scale = clip(v1, v1_min, v1_max);
+    v2 *= v1_scale;
+    const float v2_scale = clip(v2, v2_min, v2_max);
+    v1 *= v2_scale;
+}
+
+void VelocityLimiter::downscaleBoth(float& v1, float& v2, const float min, const float max) const {
+    downscaleBoth(v1, min, max, v2, min, max);
+}
+
+float VelocityLimiter::clip(float& val, const float min, const float max) const {
+    if (val > max) {
```

Comment:
what if val is 0.0 in release mode? I would not rely on assert for making sure the logic is followed. Its better to explicitly return 1 if val is near 0.0/